### PR TITLE
feat(services-user-notification): send email to user delegations

### PR DESCRIPTION
**Pull request by Herdismaria**
[Delegation lookup and add to queue](https://www.notion.so/Delegation-lookup-and-add-to-queue-f0c3235d56494dd79dc939fd7f81e7a4?pvs=4)
## What

Add logic to `notificationWorkerService` that looks up delegations for Hnipp recipient and sends the same message to delegation holders.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Added import for `DelegationsApi` in the `notificationsWorker.service.ts` file.
	- Updated email sending logic in the `notificationsWorker.service.ts` file.
	- Added a unique constraint to the `messageId` property in the `Notification` class.
	- Added a new feature flag `shouldSendEmailNotificationsToDelegations` in the `Features` enum.

- **Refactor**
	- Refactored logic for handling notifications, including delegation processing and message queueing.

- **Tests**
	- No changes.

### CODE REVIEW

**Comment by busla** [comment]
testing an AI review bot, feel free to dismiss the requested changes.

File: apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+      recipient,
+      templateId: mockHnippTemplate.templateId,
+      args: [{ key: 'organization', value: 'Test Crew' }],
+    })
+
+    // give the worker some time to process the message
+    await wait(2)
+  }
+
+  it('should send email and push notification to recipient and to delegation holders', async () => {
```
Comment (saevarma) [inline apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.spec.ts:121]: Might be missing a test case to explicitly test that a delegated notification does not continue to propagate further. The called times assert on L#146 would do it if the delegated recipient also has delegation relationship in the test setup, but might be clearer to have a specific test case for it.

File: apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.ts [hunk 2]
```diff
@@ -320,10 +330,54 @@ export class NotificationsWorkerService implements OnApplicationBootstrap {
...
+              this.logger.error('Error adding delegations to message queue', {
+                error,
+              })
+            }
+          }
+
+          // Only send push notifications for the main recipient
+          notificationPromises.push(
+            void this.handleDocumentNotification(handleNotificationArgs),
+          )
```
Comment (saevarma) [inline apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.ts:377]: ~ I would move this to the top of the `if (!message.onBehalfOf)` statement to fire this immediately before doing the delegation lookup.

File: apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.ts [hunk 3]
```diff
@@ -320,10 +330,54 @@ export class NotificationsWorkerService implements OnApplicationBootstrap {
...
         }
 
-        await Promise.all([
-          this.handleDocumentNotification(handleNotificationArgs),
-          this.handleEmailNotification(handleNotificationArgs),
-        ])
+        // should always send email notification
+        const notificationPromises = [
+          void this.handleEmailNotification(handleNotificationArgs),
+        ]
```
Comment (saevarma) [inline apps/services/user-notification/src/app/modules/notifications/notificationsWorker/notificationsWorker.service.ts:336]: `void` makes the return value of `this.handleEmailNotification` become `undefined`, making the implicit type of `notificationPromises` array of undefined, which is not correct type. I guess you are using `void` to get rid of IDE warning about not awaited promise call. You should rather explicitly type the `notificationPromises` array, which gets rid of the warnings.

```suggestion
        const notificationPromises: Promise<void>[]  = [
          this.handleEmailNotification(handleNotificationArgs),
        ]
```

**Review by saevarma — APPROVED** [review]
Great job 🚀

**Review by brynjarorng — APPROVED** [review]
LGTM

**Review by rafnarnason — APPROVED** [review]
LGTM

**Pull request merged**